x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/profile/TransportActivateProfileAction.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.profile;
+
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.action.GrantApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileAction;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileRequest;
+import org.elasticsearch.xpack.core.security.action.profile.ActivateProfileResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.support.UsernamePasswordToken;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.BearerToken;
+import org.elasticsearch.xpack.security.profile.ProfileService;
+
+public class TransportActivateProfileAction extends HandledTransportAction<ActivateProfileRequest, ActivateProfileResponse> {
+
+    private final ProfileService profileService;
+    private final AuthenticationService authenticationService;
+    private final ThreadContext threadContext;
+
+    @Inject
+    public TransportActivateProfileAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ProfileService profileService,
+        AuthenticationService authenticationService,
+        ThreadPool threadPool
+    ) {
+        super(ActivateProfileAction.NAME, transportService, actionFilters, ActivateProfileRequest::new);
+        this.profileService = profileService;
+        this.authenticationService = authenticationService;
+        this.threadContext = threadPool.getThreadContext();
+    }
+
+    @Override
+    protected void doExecute(Task task, ActivateProfileRequest request, ActionListener<ActivateProfileResponse> listener) {
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            resolveAuthentication(
+                request.getGrant(),
+                request,
+                ActionListener.wrap(
+                    authentication -> profileService.activateProfile(authentication, listener.map(ActivateProfileResponse::new)),
```

Comment:
I would avoid passing in the `authentication` as a parameter, instead you could make the service only activate the profile of the currently authenticated user. Otherwise there is the potential for ambiguity, when the two authentications diverge.

Reply:
This is only possible if we replace `authentication` with `SecurityContext#executeWithAuthentication`. As commented above, I am not entirely convinced about this approach. Replacing authentication object also incur additional serialisation cost which I am not fond of either. We have many methods that accept `Authentication` as a parameter, e.g. methods of `TokenService`, `ApiKeyService`, `CompositeRolesStore` etc. Given callers of these methods are well controlled, I am not too worried about the confusion. I can also add some comment to help clarify the usage.